code/modules/paperwork/faxmachine.dm

```diff
@@ -1,235 +1,675 @@
+
+	LAZYSET(., "quick_dial_targets", uiqd)
+	LAZYSET(., "fax_history",        fax_history)
+	LAZYSET(., "scanner_item",       "[scanner_item]")
+	LAZYSET(., "is_cooling_down",    (time_cooldown_end > world.timeofday))
+	LAZYSET(., "dest_uri",           dest_uri)
+	LAZYSET(., "src",                "\ref[src]")
+	LAZYSET(., "emagged",            emagged)
+	LAZYSET(., "is_operational",     operable())
+	LAZYSET(., "page",               current_page)
+
+	//Printer stuff
+	if(printer)
+		LAZYADD(., printer.ui_data(user))
+
+	//Card reader stuff
+	var/obj/item/card/C = card_reader?.get_card()
+	LAZYSET(., "has_card_reader",  !isnull(card_reader))
+	LAZYSET(., "id_card",          C)
+	LAZYSET(., "data_card",        card_reader?.get_data_card())
+	if(C)
+		var/info = C.name
+		if(istype(C, /obj/item/card/id))
+			var/obj/item/card/id/ID = C
+			info = ID.get_display_name()
+		LAZYSET(., "id_card_info", info)
+
+	//Disk stuff
+	var/obj/item/disk/D = disk_reader?.get_disk()
+	LAZYSET(., "has_disk_drive",   !isnull(disk_reader))
+	LAZYSET(., "disk",             D)
+	LAZYSET(., "disk_name",        D?.name)
+	LAZYSET(., "disk_has_qd",      D?.contains_file_type("FQD")) //If the disk has a quick dial file
+	LAZYSET(., "disk_has_file",    (D?.free_blocks < D?.block_capacity))
+
+/obj/machinery/faxmachine/ui_interact(mob/user, ui_key, datum/nanoui/ui, force_open, datum/nanoui/master_ui, datum/topic_state/state)
+	var/list/data = ui_data(user, ui_key)
+	ui = SSnano.try_update_ui(user, src, ui_key, ui, data, force_open)
+	if (!ui)
+		ui = new(user, src, ui_key, "fax_machine.tmpl", name, 640, 480)
+		ui.add_template("net_shared",                 "network_shared.tmpl") //Shared network UI stuff
+		ui.add_template("stock_parts_printer_shared", "stock_parts_printer.tmpl")
+		ui.set_initial_data(data)
+		ui.open()
+
+/obj/machinery/faxmachine/OnTopic(mob/user, list/href_list, datum/topic_state/state)
+	if(!CanInteract(user, state))
+		to_chat(user, SPAN_WARNING("You must be close to \the [src] to do this!"))
+		return TOPIC_NOACTION
+
+	if(href_list["change_page"])
+		var/pagename = href_list["change_page"]
+		current_page = pagename
+		return TOPIC_REFRESH
 
-		if(copyitem)
-			dat += "<a href='byond://?src=\ref[src];remove=1'>Remove Item</a><br><br>"
+	if(href_list["network_settings"])
+		var/datum/extension/network_device/fax/net = get_extension(src, /datum/extension/network_device)
+		net.ui_interact(user)
+		return TOPIC_REFRESH
 
-			if(sendcooldown)
-				dat += "<b>Transmitter arrays realigning. Please stand by.</b><br>"
+	// --- Sending Fax ---
+	if(href_list["send"])
+		if(!scanner_item)
+			return TOPIC_NOACTION
+		if(world.timeofday < time_cooldown_end)
+			to_chat(user, SPAN_WARNING("\The [src] isn't ready yet for sending again! [max(0, time_cooldown_end - world.timeofday) / (1 SECOND)] second\s left."))
+			return TOPIC_REFRESH
+
+		//Prioritize the quick dial value
+		dest_uri = replacetext(sanitize(href_list["quick_dial"]), " ", "_")
```

Comment:
Not sure what you mean by that? Sanitize just gets rid of all sort of crap character a user might input?

Reply:
Sanitize by default html-encodes the input. This is often correct (e.g. for input from an actual `input` call), but if it's not it can lead to double-encoding and player-facing %number stuff. I can't tell if it's correct here or not. There is a `sanitize` argument for not encoding if that's necessary.

Reply:
I'm going to take a look.